code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/ContentCardUIProvider.kt

```diff
@@ -0,0 +1,149 @@
+    }
+
+    private val _contentFlow = MutableStateFlow<List<AepUITemplate>>(emptyList())
+    private val contentFlow: StateFlow<List<AepUITemplate>> = _contentFlow
+
+    private val aepUiFlow = _contentFlow.map { templateList ->
+        templateList.mapNotNull { item -> getAepUI(item) }
+    }
+
+    /**
+     * Retrieves a flow of AepUI instances for the given surface.
+     *
+     * This function initiates the content fetch using [getContent] and then returns a flow of
+     * [AepUI] instances that represent the UI templates. The flow emits updates whenever new
+     * content is fetched or any changes occur.
+     *
+     * @return A [Flow] that emits a list of [AepUI] instances.
+     */
+    suspend fun getAepUi(): Flow<List<AepUI<*, *>>> {
+        getContent()
+        return aepUiFlow
+    }
+
+    /**
+     * Clears the current content and re-fetches new content for the given surface.
+     */
+    override suspend fun refreshContent() {
+        getContent()
+    }
+
+    /**
+     * Retrieves a flow of AepUITemplate lists for the given surface.
+     * The flow emits updates whenever new content is fetched.
+     *
+     * @return A flow that emits lists of AepUITemplate.
+     */
+    override suspend fun getContent(): Flow<List<AepUITemplate>> {
+        getAepUITemplateList { it ->
+            it.onSuccess { templateList ->
+                _contentFlow.value = templateList
+            }
+            it.onFailure { error ->
+                Log.error(
+                    MessagingConstants.LOG_TAG, SELF_TAG,
+                    "Failed to get content: ${error.message}"
+                )
+            }
+        }
+        return contentFlow
+    }
+
+    private suspend fun getAepUITemplateList(
+        completion: (Result<List<AepUITemplate>>) -> Unit
+    ) {
+        val surfaceList = mutableListOf<Surface>(surface)
+        // Retrieve propositions for the provided surface
+        Messaging.getPropositionsForSurfaces(
+            surfaceList,
+            object :
+                AdobeCallbackWithError<Map<Surface, List<Proposition>>> {
+                override fun call(resultMap: Map<Surface, List<Proposition>>?) {
+                    if (resultMap == null) {
+                        completion(
+                            Result.failure(
+                                Throwable(
+                                    "resultMap null for surfaces ${surfaceList.joinToString(",")}"
+                                )
+                            )
+                        )
+                        return
+                    }
+
+                    val templateModelList = mutableListOf<AepUITemplate>()
+                    Log.debug(
+                        MessagingConstants.LOG_TAG,
+                        SELF_TAG,
+                        "getPropositionsForSurfaces callback contained Null Map"
+                    )
+
+                    for ((_, propositions) in resultMap.entries) {
```

Comment:
nit: Do we need the nested for loop here? Does `val propositions: resultMap[surface]` followed by for-in not work?

Reply:
Done